wrappers/ntdsapi.go

```diff
@@ -0,0 +1,151 @@
+ */
+
+package wrappers
+
+import (
+	"syscall"
+	"unsafe"
+)
+
+const (
+	DS_SPN_DNS_HOST  = 0
+	DS_SPN_DN_HOST   = 1
+	DS_SPN_NB_HOST   = 2
+	DS_SPN_DOMAIN    = 3
+	DS_SPN_NB_DOMAIN = 4
+	DS_SPN_SERVICE   = 5
+)
+
+const (
+	DS_SPN_ADD_SPN_OP     = 0
+	DS_SPN_REPLACE_SPN_OP = 1
+	DS_SPN_DELETE_SPN_OP  = 2
+)
+
+var (
+	modntdsapi = syscall.NewLazyDLL("ntdsapi.dll")
+
+	procDsBindW              = modntdsapi.NewProc("DsBindW")
+	procDsFreeSpnArrayW      = modntdsapi.NewProc("DsFreeSpnArrayW")
+	procDsGetSpnW            = modntdsapi.NewProc("DsGetSpnW")
+	procDsMakeSpnW           = modntdsapi.NewProc("DsMakeSpnW")
+	procDsServerRegisterSpnW = modntdsapi.NewProc("DsServerRegisterSpnW")
+	procDsUnBindW            = modntdsapi.NewProc("DsUnBindW")
+	procDsWriteAccountSpnW   = modntdsapi.NewProc("DsWriteAccountSpnW")
+)
+
+func DsBind(domainControllerName *uint16, dnsDomainName *uint16, hDS *syscall.Handle) error {
+	r1, _, _ := syscall.Syscall(
+		procDsBindW.Addr(),
+		3,
+		uintptr(unsafe.Pointer(domainControllerName)),
+		uintptr(unsafe.Pointer(dnsDomainName)),
+		uintptr(unsafe.Pointer(hDS)))
+	if err := syscall.Errno(r1); err != ERROR_SUCCESS {
+		return err
+	}
+	return nil
+}
+
+func DsFreeSpnArray(cSpn uint32, spn **uint16) {
+	syscall.Syscall(
+		procDsFreeSpnArrayW.Addr(),
+		2,
+		uintptr(cSpn),
+		uintptr(unsafe.Pointer(spn)),
+		0)
+}
+
+func DsGetSpn(serviceType int32, serviceClass *uint16, serviceName *uint16, instancePort uint16, cInstanceNames uint16, instanceNames **uint16, instancePorts *uint16, cSpn *uint32, spn ***uint16) error {
+	r1, _, _ := syscall.Syscall9(
+		procDsGetSpnW.Addr(),
+		9,
+		uintptr(serviceType),
+		uintptr(unsafe.Pointer(serviceClass)),
+		uintptr(unsafe.Pointer(serviceName)),
+		uintptr(instancePort),
+		uintptr(cInstanceNames),
+		uintptr(unsafe.Pointer(instanceNames)),
+		uintptr(unsafe.Pointer(instancePorts)),
+		uintptr(unsafe.Pointer(cSpn)),
+		uintptr(unsafe.Pointer(spn)))
+	if err := syscall.Errno(r1); err != ERROR_SUCCESS {
+		return err
+	}
+	return nil
+}
+
+func DsMakeSpn(serviceClass *uint16, serviceName *uint16, instanceName *uint16, instancePort uint16, referrer *uint16, spnLength *uint32, spn *uint16) error {
+	r1, _, _ := syscall.Syscall9(
+		procDsMakeSpnW.Addr(),
+		7,
+		uintptr(unsafe.Pointer(serviceClass)),
+		uintptr(unsafe.Pointer(serviceName)),
+		uintptr(unsafe.Pointer(instanceName)),
+		uintptr(instancePort),
+		uintptr(unsafe.Pointer(referrer)),
+		uintptr(unsafe.Pointer(spnLength)),
+		uintptr(unsafe.Pointer(spn)),
+		0,
+		0)
+	if err := syscall.Errno(r1); err != ERROR_SUCCESS {
+		return err
+	}
+	return nil
+}
+
+func DsServerRegisterSpn(operation int32, serviceClass *uint16, userObjectDN *uint16) error {
+	r1, _, _ := syscall.Syscall(
+		procDsServerRegisterSpnW.Addr(),
+		3,
+		uintptr(operation),
+		uintptr(unsafe.Pointer(serviceClass)),
+		uintptr(unsafe.Pointer(userObjectDN)))
+	if err := syscall.Errno(r1); err != ERROR_SUCCESS {
+		return err
+	}
+	return nil
+}
```

Comment:
It turns out that this one function does everything we need to register a service principal name, but I didn't realize it existed until I had written the wrappers for the other functions in this file.  This is why they're here but unused.  I figured I'll leave them in place in case we need them later.